cloudevents/extensions/attrschema.md

```diff
@@ -0,0 +1,156 @@
+how to interact with them.
+
+Although JSON Schema is used to validate JSON Documents, Attrschema provides an
+algorithm to validate non-JSON formatted CloudEvents (XML, Protobuf, etc.) as well.
+
+The Attrschema meta-schema is defined under [attrschema-schema.yaml](attrschema-schema.yaml)
+
+### Validation Algorithm
+To validate a given CloudEvent the attrschema implementation MUST:
+
+1. Create a map of all attributes of a given event, where the keys are the attribute
+    names and the values are the attribute values
+2. Strip all `null` OPTIONAL attributes from the event if present
+3. All assumed values MUST be evaluated as-if the event was translated to another
+   format. (`datacontenttype` assumptions of the JSON format for example)
+4. Map all attribute values into the string canonical string representation as
+   defined in the [CloudEvent spec](../spec.md#type-system) corresponding to 
+   the attribute type.
+5. Validate the given dictionary using Attrschema document and the rules defined by the
+    [JSON Schema Spec][json-schema-spec] 
+6. Attributes are considered adhering to the attrschema document if the JSON schema
+   validation succeeded.
+
+Note: in addition to the attribute schema, the usual CloudEvent attribute constraints
+still apply, even though they are not expressed in the schema explicitly. 
+
+### Type Annotation
+Attrschema provides a way to annotate the defined attributes with the CloudEvent type
+information using the `cetype` keyword.
+
+This keyword indicates which of the standard CloudEvent types this attribute holds.
+
+This keyword MUST be one of `boolean`, `integer`, `string`, `binary`, `uri`,
+`uri-reference`, or `timestamp`.
+
+When `uri` is used the `format` keyword MUST be set to `uri`
+When `uri-reference` is used the `format` keyword MUST be set to `uri-reference`
+When `timestamp` is used the `format` keyword MUST be set to `date-time`
+
+#### Examples
+Here is an example attrschema document
+```json
+{
+  "properties": {
+    "id": {
+      "type": "string",
+      "pattern": "^A[0-9]{3}-[0-9]{4}-[0-9]{4}$"
+    },
+    "myattr": {
+      "type": "string"
+    },
+    "datacontenttype": {
+      "type": "string",
+      "enum": [
+         "text/xml",
+         "application/json"
+      ],
+    }
+  },
+  "required": ["myattr"]
+}
+```
+##### Json Event
+```json
+{
+    "specversion" : "1.0",
+    "type" : "com.github.pull_request.opened",
+    "source" : "https://github.com/cloudevents/spec/pull",
+    "id" : "A234-1234-1234",
+    "time" : "2018-04-05T17:31:00Z",
+    "myattr" : 5,
+    "yourattr" : true ,
+    "datacontenttype" : "text/xml",
+    "data" : "<much wow=\"xml\"/>"
+}
+```
+Notice that this event is adhering to the example attrschema even though the
+`myattr` is of type `integer` and not `string`. The reason for it is
+that the `myattr` value MUST be converted to the string representation
+before validation. 
```

Comment:
While I understand the idea of converting everything to a string for the purposes of the check, it feels odd that "myattr" isn't defined as an int. It seems like we should check its "type" as part of the original format (if the format support types), and then check to see if the value is syntactically correct per the pattern/enum/...